### ISSUE TITLE
BERT LOSS FUNCTION

### ISSUE THREAD

**Issue by alshahrani2030** [huggingface/transformers#2643]
My question is that can I use KLDivLoss instead of CrossEntropyLoss when I fine-tune BERT for classification? the reason for that is that I want to pass the weight of each class(e.g for binary classification, instead of 1 or 0 I will pass the probability distribution )

Thank you in advance

**Comment by cronoik** [comment]
Sure you can do that. Create a class which inherits from [BertForSequenceClassification](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_bert.py#L1122) and overwrite the [forward](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_bert.py#L1134) method.

**Comment by LysandreJik** [comment]
Instead of overwriting the forward method you can retrieve the hidden states and compute the loss as you would do with any PyTorch model. 

The loss is only computed by the model when you hand the `labels` to the model, which is not a required argument.

**Comment by paul-you** [comment]
Hi, 
I tried implementing that for _BertForSequenceClassification_ like this: 

```
golden_logits = …

outputs = model(**inputs)

# the logits from BERT
logits = outputs[1]

loss_fct = KLDivLoss(reduction='sum')
loss = loss_fct(F.log_softmax(logits, dim=-1), F.softmax(golden_logits, dim=-1))

```
Am I missing something ?

**Comment by Hemantr05** [comment]
> Instead of overwriting the forward method you can retrieve the hidden states and compute the loss as you would do with any PyTorch model.
> 
> The loss is only computed by the model when you hand the `labels` to the model, which is not a required argument.

Could you please elaborate on the same please?

**Comment by LysandreJik** [comment]
What do you want me to elaborate on?

**Comment by Hemantr05** [comment]
> Instead of overwriting the forward method you can retrieve the hidden states and compute the loss as you would do with any PyTorch model.
This @LysandreJik

**Comment by alshahrani2030** [comment]
> Instead of overwriting the forward method you can retrieve the hidden states and compute the loss as you would do with any PyTorch model.
> The loss is only computed by the model when you hand the labels to the model, which is not a required argument.

I have two classes in my data and I can calculate the probability of each class from given scores. After that, do you mean passing the probability of each class and the loss return by the model??
Thank you in advance

**Comment by alshahrani2030** [comment]
I got this error when I called 
loss.backward()
 loss is "torch.float64" type


![error-KLDiv](https://user-images.githubusercontent.com/55197626/81520607-c7436a80-9312-11ea-91ae-85fb7f032d96.PNG)

**Comment by Mahhos** [comment]
> My question is that can I use KLDivLoss instead of CrossEntropyLoss when I fine-tune BERT for classification? the reason for that is that I want to pass the weight of each class(e.g for binary classification, instead of 1 or 0 I will pass the probability distribution )
> 
> Thank you in advance

Hi did you manage to do this? I also need to pass class probability distribution instead of the labels and am not sure how to do this.

**Comment by drevicko** [comment]
BertForSequenceClassification.forward() returns the logits also. You can use these in any pytorch loss function (eg: KLDivLoss, not sure if you'll need to softmax them first) and then run backward on the resulting loss. It's a bit redundant (since BertForSequenceClassification's loss is still calculated), but works.

**Comment by drevicko** [comment]
> Sure you can do that. Create a class which inherits from [BertForSequenceClassification](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_bert.py#L1122) and overwrite the [forward](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_bert.py#L1134) method.

That link to the forward function is stale, and to be honest, finding the forward function is actually not trivial (there are layers of abstractions before you hit a base class --- in my IDE I end up in some dummy functions!).
